File: wows/src/main/java/com/half/wowsca/ui/encyclopedia/tabs/FlagsFragment.java
package com.half.wowsca.ui.encyclopedia.tabs;

import android.graphics.PorterDuff;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;
import androidx.core.util.Pair;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import com.half.wowsca.CAApp;
import com.half.wowsca.R;
import com.half.wowsca.alerts.Alert;
import com.half.wowsca.model.encyclopedia.holders.ExteriorHolder;
import com.half.wowsca.model.encyclopedia.items.ExteriorItem;
import com.half.wowsca.model.FlagClickedEvent;
import com.half.wowsca.ui.CAFragment;
import com.half.wowsca.ui.adapter.FlagsAdapter;

import org.greenrobot.eventbus.Subscribe;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.Iterator;
import java.util.List;

/**
 * Created by slai4 on 4/25/2016.
 */
public class FlagsFragment extends CAFragment {

    RecyclerView recyclerView;
    GridLayoutManager layoutManager;
    FlagsAdapter adapter;

    @Nullable
    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View view = inflater.inflate(R.layout.view_recycler_view, container, false);
        recyclerView = (RecyclerView) view;
        return view;
    }

    @Override
    public void onResume() {
        super.onResume();
        CAApp.getEventBus().register(this);
        ExteriorHolder holder = CAApp.getInfoManager().getExteriorItems(getContext());
        if (holder != null && holder.getItems() != null && recyclerView.getAdapter() == null) {
            layoutManager = new GridLayoutManager(getContext(), getResources().getInteger(R.integer.shipopedia_upgrade_grid));
            layoutManager.setOrientation(GridLayoutManager.VERTICAL);
            recyclerView.setLayoutManager(layoutManager);

            List<ExteriorItem> flags = new ArrayList<>();
            flags.addAll(holder.getItems().values());

            Collections.sort(flags, new Comparator<ExteriorItem>() {
                @Override
                public int compare(ExteriorItem lhs, ExteriorItem rhs) {
                    return lhs.getName().compareToIgnoreCase(rhs.getName());
                }
            });
            Collections.sort(flags, new Comparator<ExteriorItem>() {
                @Override
                public int compare(ExteriorItem lhs, ExteriorItem rhs) {
                    return rhs.type.compareToIgnoreCase(lhs.type);
                }
            });

            adapter = new FlagsAdapter(flags, getContext());
            recyclerView.setAdapter(adapter);
        }
    }

    @Override
    public void onPause() {
        super.onPause();
        CAApp.getEventBus().unregister(this);
    }

    @Subscribe
    public void flagClickedEvent(FlagClickedEvent event) {
        ExteriorHolder holder = CAApp.getInfoManager().getExteriorItems(getContext());
        ExteriorItem item = holder.get(event.id);
        if (item != null) {
            StringBuilder sb = new StringBuilder();
            sb.append(item.description);
            if (item.coef != null && !item.coef.isEmpty()) {
                sb.append("\n\n");
                Iterator<String> iter = item.coef.keySet().iterator();
                while (iter.hasNext()) {
                    String key = iter.next();
                    Pair<String, Float> pair = item.coef.get(key);
                    sb.append(pair.first);
                    if (iter.hasNext()) {
                        sb.append("\n");
                    }
                }
            }
            Drawable d = ContextCompat.getDrawable(getContext(), R.drawable.ic_flags);
            if (!CAApp.isDarkTheme(getContext()))
                d.setColorFilter(ContextCompat.getColor(getContext(), R.color.top_background), PorterDuff.Mode.MULTIPLY);
            Alert.createGeneralAlert(getActivity(), item.getName(), sb.toString(), getString(R.string.dismiss), d);
        } else {
            Toast.makeText(getContext(), R.string.resources_error, Toast.LENGTH_SHORT).show();
        }

    }
}

File: wows/src/main/java/com/half/wowsca/ui/encyclopedia/tabs/FlagsFragment.kt
package com.half.wowsca.ui.encyclopedia.tabs

import android.graphics.PorterDuff
import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.half.wowsca.CAApp.Companion.eventBus
import com.half.wowsca.CAApp.Companion.infoManager
import com.half.wowsca.CAApp.Companion.isDarkTheme
import com.half.wowsca.R
import com.half.wowsca.alerts.Alert.createGeneralAlert
import com.half.wowsca.model.FlagClickedEvent
import com.half.wowsca.model.encyclopedia.items.ExteriorItem
import com.half.wowsca.ui.CAFragment
import com.half.wowsca.ui.adapter.FlagsAdapter
import org.greenrobot.eventbus.Subscribe
import java.util.Collections

/**
 * Created by slai4 on 4/25/2016.
 */
class FlagsFragment : CAFragment() {
    var recyclerView: RecyclerView? = null
    var layoutManager: GridLayoutManager? = null
    var adapter: FlagsAdapter? = null

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View {
        val view = inflater.inflate(R.layout.view_recycler_view, container, false)
        recyclerView = view as RecyclerView
        return view
    }

    override fun onResume() {
        super.onResume()
        eventBus.register(this)
        val holder = infoManager!!.getExteriorItems(requireContext())
        if (holder.items != null && recyclerView!!.adapter == null) {
            layoutManager =
                GridLayoutManager(context, resources.getInteger(R.integer.shipopedia_upgrade_grid))
            layoutManager!!.orientation = GridLayoutManager.VERTICAL
            recyclerView!!.layoutManager = layoutManager

            val flags = holder.items?.values?.toMutableList() ?: mutableListOf()

            flags.sortWith { lhs, rhs ->
                lhs?.name?.compareTo(
                    rhs?.name ?: "",
                    ignoreCase = true
                ) ?: 0
            }
            flags.sortWith { lhs, rhs ->
                rhs?.type?.compareTo(
                    lhs?.type ?: "",
                    ignoreCase = true
                ) ?: 0
            }

            adapter = FlagsAdapter(flags.filterNotNull(), requireContext())
            recyclerView!!.adapter = adapter
        }
    }

    override fun onPause() {
        super.onPause()
        eventBus.unregister(this)
    }

    @Subscribe
    fun flagClickedEvent(event: FlagClickedEvent) {
        val holder = infoManager!!.getExteriorItems(requireContext())
        val item = holder[event.id]
        if (item != null) {
            val sb = StringBuilder()
            sb.append(item.description)
            if (item.coef != null && !item.coef!!.isEmpty()) {
                sb.append("\n\n")
                val iter: MutableIterator<String?> = item.coef!!.keys.iterator()
                while (iter.hasNext()) {
                    val key = iter.next()
                    val pair = item.coef!![key]
                    sb.append(pair!!.first)
                    if (iter.hasNext()) {
                        sb.append("\n")
                    }
                }
            }
            val d = ContextCompat.getDrawable(requireContext(), R.drawable.ic_flags)
            if (!isDarkTheme(context)) d!!.setColorFilter(
                ContextCompat.getColor(
                    requireContext(),
                    R.color.top_background
                ), PorterDuff.Mode.MULTIPLY
            )
            createGeneralAlert(activity, item.name, sb.toString(), getString(R.string.dismiss), d)
        } else {
            Toast.makeText(context, R.string.resources_error, Toast.LENGTH_SHORT).show()
        }
    }
}